### PR TITLE
Make generated migration scripts more robust if queueing one migration fails

### PR DESCRIPTION
**Pull request by timrogers**
Customers often use migration scripts generated with `gh gei generate-script` to queue up many migrations.

If queueing up one migration fails, then at the moment, the entire script explodes. This most often happens due to a transient network error when talking to GitHub.com or because archive generation fails in GitHub Enterprise Server.

This is a pretty painful experience, since customers will often leave the script running unattended (e.g. overnight), so they won't know when it terminates and they won't be able to intervene.

This changes the behaviour of our generated scripts so that, if one migration cannot be queued, we print out an error and move on to the next one.

If we have failed to queue up a migration, if applicable, it will be included in the "failed migrations" count printed at the end of the script and we will not attempt to download logs if the `--download-migration-logs` option was selected.

_I haven't written tests for this at the moment. Before I do that, I just wanted to get a review of my approach, since I wrote this code in a bit of a vacuum on a plane._

Fixes https://github.com/github/gh-gei/issues/781.

- [ ] Did you write/update appropriate tests
- [ ] Release notes updated (if appropriate)
- [ ] Appropriate logging output
- [ ] Issue linked
- [ ] Docs updated (or issue created)

### CODE REVIEW

File: src/gei/Handlers/GenerateScriptCommandHandler.cs [hunk 1]
```diff
@@ -338,6 +340,8 @@ private string GenerateSequentialAdoScript(IDictionary<string, IEnumerable<strin
         content.AppendLine(VersionComment);
         content.AppendLine(EXEC_FUNCTION_BLOCK);
 
+        content.AppendLine("$RepoMigrations = [ordered]@{}");
```
Comment (timrogers) [inline src/gei/Handlers/GenerateScriptCommandHandler.cs:343]: Recording this allows us to know what migrations generated an ID, and therefore what migrations we should wait for.

File: src/gei/Handlers/GenerateScriptCommandHandler.cs [hunk 2]
```diff
@@ -481,11 +485,13 @@ private string GetGhesRepoOptions(string ghesApiUrl, string awsBucketName, bool
         return $"--ghes-api-url \"{ghesApiUrl}\"{(awsBucketName.HasValue() ? $" --aws-bucket-name \"{awsBucketName}\"" : "")}{(noSslVerify ? " --no-ssl-verify" : string.Empty)}";
     }
 
-    private string WaitForMigrationScript(string repoMigrationKey = null) => $"gh gei wait-for-migration --migration-id $RepoMigrations[\"{repoMigrationKey}\"]";
+    private string WaitForMigrationScript(string repoMigrationKey = null) {
+        return $"if ($RepoMigrations[\"{repoMigrationKey}\"]) {{ gh gei wait-for-migration --migration-id $RepoMigrations[\"{repoMigrationKey}\"]\nif ($lastexitcode -eq 0) {{ $Succeeded++ }} else {{ $global:failed++ }} }}";
```
Comment (timrogers) [inline src/gei/Handlers/GenerateScriptCommandHandler.cs:489]: There is probably a better way to write this as a multi-line string

File: src/gei/Handlers/GenerateScriptCommandHandler.cs [hunk 3]
```diff
@@ -481,11 +485,13 @@ private string GetGhesRepoOptions(string ghesApiUrl, string awsBucketName, bool
...
+    }
 
     private string DownloadMigrationLogScript(string githubTargetOrg, string targetRepo)
     {
-        return $"gh gei download-logs --github-target-org \"{githubTargetOrg}\" --target-repo \"{targetRepo}\"";
+        return $"if ($RepoMigrations[\"{targetRepo}\"]) {{ gh gei download-logs --github-target-org \"{githubTargetOrg}\" --target-repo \"{targetRepo}\" }}";
```
Comment (timrogers) [inline src/gei/Handlers/GenerateScriptCommandHandler.cs:494]: We only want to download the logs if a migration ID was recorded for the repo. If no migration started, we can skip it.

File: src/gei/Handlers/GenerateScriptCommandHandler.cs [hunk 4]
```diff
@@ -422,7 +427,6 @@ private string GenerateParallelAdoScript(IDictionary<string, IEnumerable<string>
             foreach (var repo in repos[teamProject].Select(r => GetGithubRepoName(teamProject, r)))
             {
                 content.AppendLine(WaitForMigrationScript(repo));
-                content.AppendLine("if ($lastexitcode -eq 0) { $Succeeded++ } else { $Failed++ }");
```
Comment (timrogers) [inline src/gei/Handlers/GenerateScriptCommandHandler.cs:425]: This has now moved inside `WaitForMigrationScript`.

**Comment by dylan-smith** [comment]
Can somebody include an example of a script before and after this change, easier to understand if we can look at the actual script instead of trying to mentally model the string concat logic.

File: src/gei/Handlers/GenerateScriptCommandHandler.cs [hunk 5]
```diff
@@ -276,7 +279,7 @@ private string GenerateParallelGithubScript(IEnumerable<string> repos, string gi
 
         content.AppendLine();
         content.AppendLine("$Succeeded = 0");
-        content.AppendLine("$Failed = 0");
+        content.AppendLine("$global:failed = 0");
```
[the remark this answers is not in the document]
Comment (ArinGhazarian) [reply src/gei/Handlers/GenerateScriptCommandHandler.cs:282]: based on what I already mentioned for `Exec` we cannot do this.

**Review by ArinGhazarian — CHANGES_REQUESTED** [review]

**Comment by ArinGhazarian** [comment]
@timrogers I am working on a new PR to address this issue for all `generate-script` commands.

**Comment by timrogers** [comment]
Closing this in favour of @ArinGhazarian's upcoming implementation.  I just wrote this during a flight for a bit of fun, so I'm not at all precious about it ❤️

**Pull request closed without merging**